### PR TITLE
Shrink Unicode tables (even more)

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This shrinks the Unicode tables further, building upon the wins in #68232 (the previous
counts differ due to an interim Unicode version update, see #69929.

The new data structure is slower by around 3x, on the benchmark of looking up every
Unicode scalar value in each data set sequentially in every data set included. Note that
for ASCII, the exposed functions on `char` optimize with direct branches, so ASCII will
retain the same performance regardless of internal optimizations (or the reverse). Also,
note that the size reduction due to the skip list (from where the performance losses come)
is around 40%, and, as a result, I believe the performance loss is acceptable, as the
routines are still quite fast. Anywhere where this is hot, should probably be using a
custom data structure anyway (e.g., a raw bitset) or something optimized for frequently
seen values, etc.

This PR updates both the bitset data structure, and introduces a new data structure
similar to a skip list. For more details, see the [main.rs] of the table generator, which
describes both. The commits mostly work individually and document size wins.

As before, this is tested on all valid chars to have the same results as nightly (and the
canonical Unicode data sets), happily, no bugs were found.

[main.rs]: https://github.com/rust-lang/rust/blob/fb4a715e18b/src/tools/unicode-table-generator/src/main.rs

Set             | Previous |  New  |  % of old  | Codepoints | Ranges |
----------------|---------:|------:|-----------:|-----------:|-------:|
Alphabetic      |     3055 |  1599 |        52% |     132875 |    695 |
Case Ignorable  |     2136 |   949 |        44% |       2413 |    410 | 
Cased           |      934 |   359 |        38% |       4286 |    141 |
Cc              |       43 |     9 |        20% |         65 |      2 | 
Grapheme Extend |     1774 |   813 |        46% |       1979 |    344 | 
Lowercase       |      985 |   867 |        88% |       2344 |    652 |
N               |     1266 |   419 |        33% |       1781 |    133 |
Uppercase       |      934 |   777 |        83% |       1911 |    643 |
White_Space     |      140 |    37 |        26% |         25 |     10 |
----------------|----------|-------|------------|------------|--------|
Total           |    11267 |  5829 |        51% |     -      |   -    |

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
r? @dtolnay perhaps?

cc @jamesmunns once more for embedded

Also cc @thomcc from https://github.com/BurntSushi/ucd-generate/issues/30

@joshtriplett also brought up in https://github.com/rust-lang/rust/pull/68232#issuecomment-574533876 that we may just want to jettison the grapheme_extend table (and the only user, which is the debug printing for chars/strs). I am personally ambivalent -- certainly, for strings which are non-ASCII, it can be a big win to avoid the extensive `\u{...}` chars in printouts. But I don't know how to get data on how common it is. Certainly we intentionally made the choice to use Grapheme_Extend; see discussion on https://github.com/rust-lang/rust/pull/49283 which I believe is the most recent extensive discussion relating to this.

**Review by dtolnay — COMMENTED** [review]
Thanks, this is terrific. I agree that these lookups are fast enough that optimizing for size is the important thing.

Is there a way to measure the impact of the more complicated lookup logic on the binary size? In this case it doesn't look like the difference would be more than ~100 bytes but it's worth taking into account next to the raw table sizes. For example if we shrink grapheme_extend by 100 bytes but grow the search code it requires by 200 bytes that would not be smart.

**Comment by joshtriplett** [comment]
This looks good to me as well.

@Mark-Simulacrum Yes, I'd still like to see that change made, to save the extra 813 bytes (plus a bit for the function using it).

**Comment by Mark-Simulacrum** [comment]
It looks like the lookup logic increased for pretty much everything -- roughly doubling in size. It's still pretty small, but for the sets which had a small delta (100-200 bytes or so) most of that was eaten up by the lookup logic; I think for most we shrunk by enough that it doesn't matter too much. In any case, for the most important one -- grapheme extend -- the new set is still 56% of the old one.

With that in mind, I think changing things here is probably not too worth it, but I'll record it again in the future if I decide to look into this some more. `nm --print-size --radix=d ./nightly | rustfilt | grep unicode | sed 's@core::unicode::unicode_data::@@'` is what I used to get the statistics for this.


Set             | Old | New | Delta |
----------------|----:|----:|------:|
Alphabetic      | 151 | 341 |   190 |
White_Space     | 146 | 275 |   129 |
Case Ignorable  | 151 | 325 |   174 |
Grapheme Extend | 151 | 325 |   174 |
N               | 141 | 341 |   200 |
Cc              |  80 |  18 |   -62 |
Cased           | 146 | 325 |   178 |
Lowercase       | 146 | 261 |   115 |
Uppercase       | 146 | 261 |   115 |

File: src/libcore/unicode/unicode_data.rs [hunk 1]
```diff
@@ -1,618 +1,547 @@
...
+) -> bool {
+    // Note that this *cannot* be past the end of the array, as the last
+    // element is greater than std::char::MAX (the largest possible needle).
+    //
+    // So, we cannot have found it (i.e. Ok(idx) + 1 != length) and the correct
+    // location cannot be past it, so Err(idx) != length either.
+    //
+    // This means that we can avoid bounds checking for the accesses below, too.
+    let last_idx =
+        match short_offset_runs.binary_search_by_key(&(needle << 11), |header| header << 11) {
```
Comment (HKalbasi) [inline src/libcore/unicode/unicode_data.rs:70]: Can someone explain why these `<< 11`s are needed?

**Pull request merged**